### ISSUE TITLE
Wrong player movement implementaion

### ISSUE THREAD

**Issue by kbariotis** [huytd/agar.io-clone#138]
I've noticed that the player gets to move when the mouse is moving. (https://github.com/huytd/agar.io-clone/blob/master/client/js/app.js#L85).

This is wrong, as if the mouse stops then the player stops. Player must keep moving while the mouse is away from him. Not while the mouse moves.

There are also a lot of guys complaining about speed. Could this be the issue?

Btw, i will make a try to fix it.:)

**Comment by hunterpl** [comment]
Hello !

I have the same problem when players connect and stop move the mouse they are just stop as well, they should moving all the time. Please fix it.

**Comment by artuross** [comment]
You're wrong, every time player moves his mouse, new position is saved, but you don't need to move your mouse all the time, only at the beginning (setting direction for the first time).

**Comment by jniles** [comment]
Anytime the mouse leaves the canvas, motion stops.  This is the desired behavior in general, but the chat input and player count are both registered as off canvas and will halt player motion.  See issue #44.

It might be better to listen the document or window object, rather than the canvas since we have so many overlays.  I'll do some tests and see if one of those works.

**Comment by kbariotis** [comment]
>but you don't need to move your mouse all the time, only at the beginning (setting direction for the first time).
@artuross This is right but how is you saying different from mine?

@jniles Didn't say anything about mouse leaving the canvas. Problem is that the player's movement is registered to the mouse's movement. This means that if the mouse stops then the player stops.

**Comment by artuross** [comment]
By 'mouse stops' I understand not changing cursor's position - and this doesn't affect player's movement.

**Comment by jniles** [comment]
@kbariotis , I cannot replicate the issue you are referring to.   On my machine, [line you linked](https://github.com/huytd/agar.io-clone/blob/master/client/js/app.js#L85) begins the game with an initial mouse movement, and the player's cell continues to move whether I am moving the mouse or not.  Is it different on your machine?  What browser/os are you using?

**Comment by hunterpl** [comment]
@jniles http://ts3bahu.pl:3000 can you test it there, how can I fix this problem with movement? default files and config there.

**Comment by jniles** [comment]
@hunterpl, I see it now.  Interesting.  I'll take a look.

**Comment by jniles** [comment]
@hunterpl, @kbariotis ,  this is indeed a confusing issue.   The server at http://ts3bahu.pl:3000 indeed only moves the player on mouse movement.  Yet locally, my player moves without new input.  `console.log()`s on both the live server and locally show that the 'mousemove' event is firing at the same time on both machines...  Strange :S.

I'll keep looking.  If anyone has a proposed explanation (or fix), I'd be grateful. :)

**Comment by hunterpl** [comment]
@jniles probably the web engine could be a problem or kernel linux with npm installed?

uname -a : 

Debian Jessie ( stable )
Linux ts3bahu 3.16.0-0.bpo.4-amd64 #1 SMP Debian 3.16.7-ckt9-3~deb8u1~bpo70+1 (2015-04-27) x86_64 GNU/Linux

Web Engine: lighttpd
lighttpd/1.4.35 (ssl) - a light and fast webserver

NPM Jessie:
1.4.21

**Comment by jniles** [comment]
@hunterpl , thank you for the info.  I can replicate it on my own server (192.99.43.113:3000) from a fresh clone.  I wonder if WebSockets behave differently over a network versus a local connection.

**Comment by hunterpl** [comment]
192.99.43.113:3000 can not connect there due of connection lost in web browser.

**Comment by aarohmankad** [comment]
[I'm also getting a similar behaviour](http://3c067f82.ngrok.io/)

**Comment by jniles** [comment]
Whoops, I had killed the server.  It's up now.

**Comment by hunterpl** [comment]
Tested and same like on mine server. By the way @aarohmankad  how did you hide the port in web browser?

**Comment by IgorAntun** [comment]
That commit should fix it. It wasn't sending any information to the server if `oldx` or `oldy` were the same as `target.y` and `target.x`.

tl;dr: If the mouse position didn't change, it wouldn't send any information to the socket server.

**Comment by artuross** [comment]
@hunterpl Port 80 is default for http protocol, so if you want to avoid users to type hostname:ip you need to set port (file server/config.yaml) to 80.

**Comment by aarohmankad** [comment]
@hunterpl I used [ngrok](http://ngrok.com/). Also, I believe we can change when we accept gameInput from on [mousemove](https://github.com/huytd/agar.io-clone/blob/master/client/js/app.js#L107) to something else.

**Comment by hunterpl** [comment]
@artuross Hi there yes I'm know about this but I have on port 80 the other website that's why I'm asking about this.

**Comment by artuross** [comment]
@hunterpl You can host multiple websites on the same IP address. It's all about you server's config files.

**Comment by IgorAntun** [comment]
I think we could move that conversation to our Gitter to avoid cluttering this Issue ;)
There we could talk a little about node + http servers haha

**Comment by aarohmankad** [comment]
Please check my [local version](http://cbf73afd.ngrok.io/), I may have the issue fixed.

**Comment by IgorAntun** [comment]
@aarohmankad This issue is closed, it has been fixed already, thanks :)

**Comment by kbariotis** [comment]
I am on a Mac with latest stable Chrome. Indeed the online version @hunterpl gave is working fine. But the local one on my laptop doesn't work that way.

**Comment by IgorAntun** [comment]
Are you sure you have the **latest** version, from after I fixed this above?
